3.0/m_restrictmsg_duration.cpp

```diff
@@ -0,0 +1,146 @@
+ * Connect Class exemption (add to any connect class block you wish to exempt from this):
+ * exemptrestrictmsg="yes"
+ */
+
+
+#include "inspircd.h"
+#include "modules/account.h"
+
+class ModuleRestrictMsgDuration : public Module
+{
+	bool blockuser;
+	bool blockchan;
+	bool exemptoper;
+	bool exemptuline;
+	bool exemptregistered;
+	bool notify;
+	time_t duration;
+
+ public:
+	void Prioritize() CXX11_OVERRIDE
+	{
+		// Go last to let filter, +R, bans, etc. act first
+		ServerInstance->Modules->SetPriority(this, PRIORITY_LAST);
+	}
+
+	void ReadConfig(ConfigStatus&) CXX11_OVERRIDE
+	{
+		ConfigTag* tag = ServerInstance->Config->ConfValue("restrictmsg_duration");
+
+		const std::string target = tag->getString("target", "both");
+		if (target != "user" && target != "chan" && target != "both")
+		{
+			throw ModuleException("Invalid \"target\" of '" + target + "' in <restrictmsg_duration>. Default of both will be used.");
+		}
+
+		blockuser = (target != "chan");
+		blockchan = (target != "user");
+		exemptoper = tag->getBool("exemptoper", true);
+		exemptuline = tag->getBool("exemptuline", true);
+		exemptregistered = tag->getBool("exemptregistered", true);
+		notify = tag->getBool("notify");
+		duration = tag->getDuration("duration", 60);
+	}
+
+	ModResult OnUserPreMessage(User* user, const MessageTarget& target, MessageDetails& details) CXX11_OVERRIDE
+	{
+		LocalUser* src = IS_LOCAL(user);
+		// Only check against non-oper local users
+		if (!src || src->IsOper())
+			return MOD_RES_PASSTHRU;
+
+		// Check their connected duration
+		if (src->signon + duration <= ServerInstance->Time())
+			return MOD_RES_PASSTHRU;
+
+		// Check for connect class exemption
+		if (src->MyClass->config->getBool("exemptrestrictmsg"))
+			return MOD_RES_PASSTHRU;
+
+		if (target.type == MessageTarget::TYPE_USER)
+		{
+			if (!blockuser)
+				return MOD_RES_PASSTHRU;
+
+			User* const dst = target.Get<User>();
+
+			// Target is Oper exemption
+			if (exemptoper && dst->IsOper())
+				return MOD_RES_PASSTHRU;
+			// Target is on a U-Lined server exemption
+			if (exemptuline && dst->server->IsULine())
+				return MOD_RES_PASSTHRU;
+		}
+		else if (target.type == MessageTarget::TYPE_CHANNEL && !blockchan)
+			return MOD_RES_PASSTHRU;
+
+		// Source is registered (and identified) exemption
+		if (exemptregistered)
+		{
+			const AccountExtItem* accountname = GetAccountExtItem();
+			if (accountname)
+			{
+				const std::string* account = accountname->get(src);
+				if (account && !account->empty())
+					return MOD_RES_PASSTHRU;
+			}
+		}
+
+		if (notify)
+		{
+			if (target.type == MessageTarget::TYPE_USER)
+			{
+				src->WriteNumeric(ERR_CANTSENDTOUSER, target.Get<User>()->nick,
+					InspIRCd::Format("You cannot send messages within the first %lu seconds of connecting.",
+					duration));
+			}
+			else
+			{
+				src->WriteNumeric(ERR_CANNOTSENDTOCHAN, target.Get<Channel>()->name,
```

Comment:
The target might be `TYPE_SERVER` here. You should update your else statement to explicitly check for TYPE_CHANNEL.